[ECS] Add container stats collection for ECS managed instances with sidecar deployment

stzou opened the pull request:
### What does this PR do?
This PR is built on top of https://github.com/DataDog/datadog-agent/pull/41712 and https://github.com/DataDog/datadog-agent/pull/42203. Read description in those PRs for full context. 

This PR extends on the agent to report `container.*` metrics when running as a side car in ECS Managed Instances. To do this, the agent retrieves the metrics from the ECS v4/task/stats endpoint which is the equivalent of the v2/stats endpoint used for fargate.

This container stats implementation is meant to provide metric support for ECS Managed Instances until daemon scheduling is supported.

### Motivation
See description here: https://github.com/DataDog/datadog-agent/pull/42203
The agent must be deployed as a sidecar with Managed Instances for now. We want to treat the agent in a similar manner as fargate where we only collect task-level metrics. To do this, we need to collect container stats metrics from the v4 metadata endpoint:
https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4-managed-instances.html

### Describe how you validated your changes
- Deployed the image `datadog/agent-dev:stzou-exp-65-full` as a sidecar on ECS Managed Instance with the env var `DD_ECS_DEPLOYMENT_MODE=sidecar`.
- confirmed that `container.*` metrics are being reported by the agent

Example of task definition used:
```
{
    "containerDefinitions": [
        {
            "cpu": 200,
            "environment": [
                {
                    "name": "DD_API_KEY",
                    "value": "..."
                },
                {
                    "name": "DD_ECS_DEPLOYMENT_MODE",
                    "value": "sidecar"
                },
                {
                    "name": "DD_PROCESS_AGENT_PROCESS_COLLECTION_ENABLED",
                    "value": "true"
                }
            ],
            "essential": true,
            "image": "datadog/agent-dev:stzou-exp-65-full",
            "logConfiguration": {
                "logDriver": "awslogs",
                "options": {
                    "awslogs-group": "/ecs/datadog-agent-task-managed-instances",
                    "awslogs-create-group": "true",
                    "awslogs-region": "us-west-2",
                    "awslogs-stream-prefix": "ecs"
                },
                "secretOptions": []
            },
            "memoryReservation": 512,
            "mountPoints": [
                {
                    "containerPath": "/var/run/datadog",
                    "readOnly": false,
                    "sourceVolume": "dd-sockets"
                }
            ],
            "name": "datadog-agent",
            "portMappings": [],
            "systemControls": [],
            "volumesFrom": []
        },
        {
            "command": [
                "/bin/sh -c \"echo '<html><head><title>Amazon ECS Sample App</title><style>body {margin-top: 40px; background-color: #333;} </style></head><body><div style=color:white;text-align:center><h1>Amazon ECS Sample App</h1><h2>Congratulations!</h2><p>Your application is now running on a container in Amazon ECS using Amazon ECS Managed Instances.</p></div></body></html>' >  /usr/local/apache2/htdocs/index.html && httpd-foreground\""
            ],
            "cpu": 412,
            "entryPoint": [
                "sh",
                "-c"
            ],
            "environment": [],
            "essential": true,
            "image": "public.ecr.aws/docker/library/httpd:latest",
            "memoryReservation": 200,
            "mountPoints": [],
            "name": "managed-instances-app",
            "portMappings": [
                {
                    "containerPort": 80,
                    "hostPort": 80,
                    "name": "managed-instances-app-80-tcp",
                    "protocol": "tcp"
                }
            ],
            "systemControls": [],
            "volumesFrom": []
        },
        {
            "cpu": 256,
            "environment": [
                {
                    "name": "DD_DOGSTATSD_URL",
                    "value": "unix:///var/run/datadog/dsd.socket"
                },
                {
                    "name": "DD_TRACE_AGENT_URL",
                    "value": "unix:///var/run/datadog/apm.socket"
                }
            ],
            "essential": false,
            "image": "ghcr.io/datadog/apps-tracegen:main",
            "memory": 512,
            "mountPoints": [
                {
                    "containerPath": "/var/run/datadog",
                    "readOnly": false,
                    "sourceVolume": "dd-sockets"
                }
            ],
            "name": "datadog-apm-app",
            "portMappings": [],
            "systemControls": [],
            "volumesFrom": []
        }
    ],
    "cpu": "1024",
    "executionRoleArn": "arn:aws:iam::12345:role/ecsTaskExecutionRole",
    "family": "datadog-agent-task-managed-instance",
    "memory": "720",
    "networkMode": "host",
    "pidMode": "task",
    "placementConstraints": [],
    "requiresCompatibilities": [
        "MANAGED_INSTANCES"
    ],
    "volumes": [
        {
            "host": {},
            "name": "dd-sockets"
        }
    ],
    "tags": []
}
```

### Additional Notes

JLineaweaver committed:
[ecs] Split launch type from deployment mode

JLineaweaver committed:
Remove unneeded import

JLineaweaver committed:
Fix deployment test

JLineaweaver committed:
Refactor the ecsfargate code within ecs

We want to refactor the fargate code to separate agent deployment
strategy vs ecs launch type. To do so we removed ecsfargate folder and
separated everything into just the ecs folder. We now share as much code
as possible but make daemon vs sidecar clear

JLineaweaver committed:
Fix unused param linting issues

JLineaweaver committed:
Fix additional linter issues

JLineaweaver committed:
Move initialization to correct call + cleanup comments

JLineaweaver committed:
Cleanup fargate codeowners

JLineaweaver committed:
Merge branch 'main' into jlineaweaver/exp-3

JLineaweaver committed:
Add release note

JLineaweaver committed:
Fix release note

JLineaweaver committed:
[ecs] Add managed instances detection

JLineaweaver committed:
Change IsFargate to IsSidecar

JLineaweaver committed:
Prevent incompatible launch type + deployment type matches

stzou committed:
Add v4 container stats collection for ECS Managed Instance with sidecar

stzou committed:
fix linter

JLineaweaver committed:
Change function name back to sidecar

JLineaweaver committed:
Add missing refactor functions

stzou committed:
set container runtime for managed instances

JLineaweaver committed:
Add release note

JLineaweaver committed:
Merge branch 'main' into jlineaweaver/exp-3

JLineaweaver committed:
Merge branch 'jlineaweaver/exp-3' into jlineaweaver/managed-instances

stzou committed:
rename isSidecarInstance

JLineaweaver committed:
Correct isSidecarMode to use correct config type

JLineaweaver committed:
Remove unneeded nil check for IsSidecarMode

zhuminyi commented on comp/core/workloadmeta/collectors/util/ecs_util.go:
```diff
@@ -86,6 +87,10 @@ func ParseV4Task(task v3or4.Task, seen map[workloadmeta.EntityID]struct{}) []wor
 		entity.LaunchType = workloadmeta.ECSLaunchTypeFargate
 		source = workloadmeta.SourceRuntime
 	}
+	if strings.ToUpper(task.LaunchType) == "MANAGED_INSTANCES" && fargate.IsFargateInstance() {
```
Same question as above.

JLineaweaver committed:
Fix metrics by adding correct tagging for MI

We need to correctly add tags in workloadmeta for managed instances. We
want to treat it like faragte if sidecar otherwise treat it like ecs on
ec2.

JLineaweaver commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser.go:
```diff
@@ -233,6 +234,8 @@ func (c *collector) parseV4TaskForSidecar(task *v3or4.Task) []workloadmeta.Colle
 		if container, ok := taskEvents[i].Entity.(*workloadmeta.Container); ok {
 			if c.actualLaunchType == workloadmeta.ECSLaunchTypeFargate {
 				container.Runtime = workloadmeta.ContainerRuntimeECSFargate
+			} else if c.actualLaunchType == workloadmeta.ECSLaunchTypeManagedInstances && fargate.IsFargateInstance() {
```
`IsFargateInstance` is just a check to see if we're running as a sidecar

JLineaweaver replied:
Wanted to rename that function, but that brings in way more reviews

stzou committed:
add init script for managed instances with sidecar mode

JLineaweaver committed:
Merge remote-tracking branch 'origin/main' into jlineaweaver/managed-instances

JLineaweaver committed:
Add back missing file from merge conflict

stzou committed:
address issues in ecs-managed init script

zhuminyi reviewed (approved):
LGTM on behalf of container platform team. But we need to rename `IsFargateInstance` as it is too confusing.

stzou committed:
add release note

JLineaweaver committed:
Merge branch 'main' into stzou/EXP-65
